### PR TITLE
Add cbc flush before statement parsing in constructors

### LINKED ISSUE #4238: Assertion context_p->last_cbc_opcode == PARSER_CBC_UNAVAILABLE in parser_save_context
###### JerryScript revision
629a0e51

###### Build platform
Linux-5.4.0-42-generic-x86_64-with-glibc2.29

###### Build steps
```
./tools/build.py --clean --debug --compile-flag=-fsanitize=address \
--compile-flag=-m32 --compile-flag=-fno-omit-frame-pointer \
--compile-flag=-fno-common --compile-flag=-g \
--strip=off --system-allocator=on --logging=on \
--linker-flag=-fuse-ld=gold --error-messages=on --profile=es2015-subset \
--builddir=build --stack-limit=16 \
```

###### Test case
```javascript
class $ {
    constructor() {
        function $() {}
    }
```

###### Output
```text
ICE: Assertion 'context_p->last_cbc_opcode == PARSER_CBC_UNAVAILABLE' failed at jerryscript/jerry-core/parser/js/js-parser.c(parser_save_context):2411.
Error: ERR_FAILED_INTERNAL_ASSERTION
AddressSanitizer:DEADLYSIGNAL
=================================================================
==2678543==ERROR: AddressSanitizer: ABRT on unknown address 0x0028df0f (pc 0xf7eeeb49 bp 0xffe8845c sp 0xffe88440 T0)
    #0 0xf7eeeb48  (linux-gate.so.1+0xb48)
    #1 0xf76e14c5 in raise (/lib/i386-linux-gnu/libc.so.6+0x354c5)
    #2 0xf76c93f6 in abort (/lib/i386-linux-gnu/libc.so.6+0x1d3f6)
    #3 0x566ea077 in jerry_port_fatal jerryscript/jerry-port/default/default-fatal.c:30
    #4 0x56604c25 in jerry_fatal jerryscript/jerry-core/jrt/jrt-fatals.c:63
    #5 0x56604c66 in jerry_assert_fail jerryscript/jerry-core/jrt/jrt-fatals.c:87
    #6 0x56614ec5 in parser_save_context jerryscript/jerry-core/parser/js/js-parser.c:2411
    #7 0x566168f8 in parser_parse_function jerryscript/jerry-core/parser/js/js-parser.c:2544
    #8 0x566d3890 in parser_parse_function_statement jerryscript/jerry-core/parser/js/js-parser-statm.c:798
    #9 0x566de723 in parser_parse_statements jerryscript/jerry-core/parser/js/js-parser-statm.c:2932
    #10 0x56616c3b in parser_parse_function jerryscript/jerry-core/parser/js/js-parser.c:2615
    #11 0x566bc7c0 in parser_parse_class_literal jerryscript/jerry-core/parser/js/js-parser-expr.c:597
    #12 0x566be747 in parser_parse_class jerryscript/jerry-core/parser/js/js-parser-expr.c:1002
    #13 0x566de6ea in parser_parse_statements jerryscript/jerry-core/parser/js/js-parser-statm.c:2911
    #14 0x5661429a in parser_parse_source jerryscript/jerry-core/parser/js/js-parser.c:2317
    #15 0x56618988 in parser_parse_script jerryscript/jerry-core/parser/js/js-parser.c:3099
    #16 0x56577c57 in jerry_parse jerryscript/jerry-core/api/jerry.c:451
    #17 0x56571881 in main jerryscript/jerry-main/main-unix.c:112
    #18 0xf76caee4 in __libc_start_main (/lib/i386-linux-gnu/libc.so.6+0x1eee4)
    #19 0x56571144 in _start (jerryscript/build/bin/jerry+0x1b144)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: ABRT (linux-gate.so.1+0xb48) 
==2678543==ABORTING
```

<sup>Found by [Fuzzinator](http://fuzzinator.readthedocs.io/) with [grammarinator](https://github.com/renatahodovan/grammarinator).</sup>

### PR DESCRIPTION
**Pull request by pmarkee**
Fixes #4238

JerryScript-DCO-1.0-Signed-off-by: Peter Marki marpeter@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/parser/js/js-parser.c [hunk 1]
```diff
@@ -2302,6 +2302,12 @@ parser_parse_function (parser_context_t *context_p, /**< context */
   }
 
   lexer_next_token (context_p);
+#if ENABLED (JERRY_ESNEXT)
+  if ((context_p->status_flags & (PARSER_CLASS_CONSTRUCTOR | PARSER_ALLOW_SUPER_CALL)) == PARSER_CLASS_CONSTRUCTOR)
+  {
+    parser_flush_cbc (context_p);
+  }
+#endif /* ENABLED (JERRY_ESNEXT) */
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser.c:2310]: There is a code above:
```
if ((context_p->status_flags & (PARSER_CLASS_CONSTRUCTOR | PARSER_ALLOW_SUPER_CALL)) == PARSER_CLASS_CONSTRUCTOR)
  {
    parser_emit_cbc_ext (context_p, CBC_EXT_RUN_FIELD_INIT);
  }
```
Maybe we could put a flush after it. But what cause the flush in a normal case (when no parse-error)?

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: jerry-core/parser/js/js-parser.c (hunk 1 above)
Comment (pmarkee) [reply jerry-core/parser/js/js-parser.c:2310]: The linked issue also occurs with correct syntax, any time a constructor begins with a function statement. No parse error happens when I am stepping through the example with `cgdb`. For example this fails:
```js
class foo  {
    constructor() {
        function bar() {}
    }
}
```
But this runs as expected:
```js
class foo  {
    constructor() {
        var a = 5
        function bar() {}
    }
}
```
In the second case, a flush occurs in `parser_emit_cbc_literal()`, that's why the `context_p->last_cbc_opcode == PARSER_CBC_UNAVAILABLE` check will pass when we begin parsing the `function bar() {}`. I didn't check for every kind of statement, but I think there must be some kind of flush in all of them.
And the tests are all passing so unless there is some corner case we don't test, I think this should have no side effects.

Comment (zherczeg) [reply jerry-core/parser/js/js-parser.c:2310]: I also asked: what about putting the flush after the CBC_EXT_RUN_FIELD_INIT?

Comment (pmarkee) [reply jerry-core/parser/js/js-parser.c:2310]: Yes, I forgot about that, sorry. We can put it there of course, I will update the patch.

**Review by rerobika — APPROVED** [review]
LGTM

**Review by zherczeg — APPROVED** [review]
LGTM
